user-guide/Advanced_Modules/Dashboards_and_Low_Code_Apps/GQI/Extensions/Ad_hoc_data/Ad_hoc_Tutorials_GQIDMS.md

```diff
@@ -0,0 +1,121 @@
+---
+uid: Ad_hoc_Tutorials_GQIDMS
+---
+
+# Building a GQI data source that retrieves data from a DMS
+
+In this tutorial, you will learn how you can create a GQI data source that retrieves the active client connections from your DMS.
+
+Expected duration: 15 minutes.
+
+> [!NOTE]
+> This tutorial uses DataMiner version 10.3.4/10.4.0 onwards.
+
+## Prerequisites
+
+- [Visual Studio](https://visualstudio.microsoft.com/downloads/) with [DataMiner Integration Studio](xref:Installing_and_configuring_DataMiner_Integration_Studio)
+
+## Overview
+
+- [Step 1: Create the ad hoc data source](#step-1-create-the-ad-hoc-data-source)
+- [Step 2: Fetch the client connections from the DMS](#step-2-fetch-the-client-connections-from-the-dms)
+- [Step 3: Parse the responses into a GQIPage](#step-3-parse-the-responses-into-a-gqipage)
+
+## Step 1: Create the ad hoc data source
+
+1. Create a new class that implements the [IGQIDatasource](xref:GQI_IGQIDataSource) interface.
+
+> [!NOTE]
+> If certain types cannot be found in the file, verify if the *Skyline.DataMiner.Dev.Automation* NuGet package has the correct version. Go to *Tools* > *NuGet Package Manager* > *Manage NuGet Packages for Solution*. Select *Skyline.DataMiner.Dev.Automation*, and verify whether the version installed for the current project is at least *10.3.4*.
+
+1. Implement the GetColumns function to define the columns that the data source will contain. Our data source will include two columns: `Client` (of type `String`) and `Connection Time` (of type `DateTime`).
+
+1. Implement the `GetNextPage` function to provide the actual data. For now, leave it empty since we still need to retrieve the data from the DMS.
+
+```csharp
+[GQIMetaData(Name = "Client connections")]
+public class ClientConnections : IGQIDataSource
+{
+    public GQIColumn[] GetColumns()
+    {
+        return new GQIColumn[]
+        {
+            new GQIStringColumn("Client"),
+            new GQIDateTimeColumn("Connection Time"),
+        };
+    }
+
+    public GQIPage GetNextPage(GetNextPageInputArgs args)
+    {
+        // TODO: Fetch the data
+        return default;
+    }
+}
+```
+
+## Step 2: Fetch the client connections from the DMS
+
+To retrieve the current client connections, communication with the DMS is required. You can use the [GQIDMS](xref:GQI_GQIDMS) object, available in the [OnInitInputArgs](xref:GQI_OnInitInputArgs) argument of the `OnInit` function, by implementing the [IGQIOnInit](xref:GQI_IGQIOnInit) interface.
+
+1. Add the [IGQIOnInit](xref:GQI_IGQIOnInit) interface to the class.
+
+1. Implement the `OnInit` function, which receives [OnInitInputArgs](xref:GQI_OnInitInputArgs) as input arguments. Store the DMS property from the arguments in the class for later use.
```

Comment:
We should call regular "input arguments" to C# methods "parameters".
Also to avoid confusion with the actual data source input arguments.

Reply:
This step should probably be split up? (see doc guidelines about avoiding steps with multiple actions)